python_modules/dagster/dagster/_core/definitions/asset_reconciliation_sensor.py

```diff
@@ -0,0 +1,370 @@
+                        # being materialized. We'll materialize the asset on the next tick when the
+                        # materialization of p is complete
+                        parent_asset_event_records = {pp: (False, 0.0) for pp in parent_assets}
+
+                        return parent_asset_event_records
+            # check if there is a completed materialization for p
+            event_records = context.instance.get_event_records(
+                EventRecordsFilter(
+                    event_type=DagsterEventType.ASSET_MATERIALIZATION,
+                    asset_key=p,
+                ),
+                ascending=False,
+                limit=1,
+            )
+
+            if event_records and event_records[0].event_log_entry.timestamp > cursor_timestamp:
+                # if the run for the materialization of p also materialized current_asset, we
+                # don't consider p "updated" when determining if current_asset should materialize
+                other_materialized_asset_records = context.instance.get_records_for_run(
+                    run_id=event_records[0].event_log_entry.run_id,
+                    of_type=DagsterEventType.ASSET_MATERIALIZATION_PLANNED,
+                ).records
+                other_materialized_assets = [
+                    event.event_log_entry.dagster_event.event_specific_data.asset_key
+                    for event in other_materialized_asset_records
+                ]
+                if current_asset in other_materialized_assets:
+                    # we still update the cursor for p so this materialization isn't considered
+                    # on the next sensor tick
+                    parent_asset_event_records[p] = (
+                        False,
+                        event_records[0].event_log_entry.timestamp,
+                    )
+                else:
+                    # current_asset was not updated along with p, so we consider p updated
+                    parent_asset_event_records[p] = (
+                        True,
+                        event_records[0].event_log_entry.timestamp,
+                    )
+            else:
+                # p has not been materialized and will not be materialized by the sensor
+                parent_asset_event_records[p] = (False, 0.0)
+
+    return parent_asset_event_records
+
+
+def _make_sensor(
+    selection: AssetSelection,
+    name: str,
+    wait_for_all_upstream: bool,
+    wait_for_in_progress_runs: bool,
+    minimum_interval_seconds: Optional[int] = None,
+    description: Optional[str] = None,
+    default_status: DefaultSensorStatus = DefaultSensorStatus.STOPPED,
+) -> MultiAssetSensorDefinition:
+    """Creates the sensor that will monitor the parents of all provided assets and determine
+    which assets should be materialized (ie their parents have been updated).
+
+    The cursor for this sensor is a dictionary mapping stringified AssetKeys to a timestamp (float). For each
```

Comment:
Why timestamp instead of storage ID? Timestamps are less reliable for ordering events because of the way computer clocks work. And I think less efficient for us to query on?

Reply:
Since the query just gets the single latest asset materialization event, i dont need to use the timestamp as part of the query, and just use it to see if the returned asset materialization event is more recent. I could switch the cursor to the most recent storage id, that would work as well.

Question about that - lets say that assets A and B are parents to C and both A and B were materialized and B has the higher storage id (lets say 3). So then we set the cursor for C to 3. On the next tick when I query for asset materialization events for A can I pass in the `after_cursor` of 3 and it'll only return the events after storage id 3 even if that storage id doesn't correspond to a materialization of A? I would assume that should be fine but the event fetching is a bit of a black box to me so i dont know if there are any gotchas.

Reply:
It would definitely work on most of our event log storages. But it might not work with run-sharded event log storages. Thoughts @clairelin135 ?

Reply:
actually, i think if i keep it so that i only fetch the latest materialization and then do a storage id comparison after (instead of passing `after_cursor`) i can cache all of these calls as well